### PR TITLE
TRI-7 Add 'clj -M:build-db migrate' command

### PR DESCRIPTION
**Pull request by rjsheperd**
## Purpose

TRI-7 is meant to add a automatic database migration command to Triangulum

This is necessary to make auto-deployment possible for projects with databases, as well as help to simplify the challenges with tracking which DB migrations have been run.

From the `triangulu.migrate/migrate!` docs:
> Performs the database migrations stored in the `src/sql/changes/` directory.
> Migrations must be stored in chronological order (e.g. `2021-02-28_add-users-table.sql`).
>
>Migrations run inside of a transaction block to ensure the entire migration is completed prior to being committed to the database.
>
>  Currently, this tool does not support rollbacks.
>
>  If a migration fails, all migrations which follow it will be cancelled.
>
>  Migrations which have been completed are stored in a table `tri.migrations`, and include a SHA-256 hash of the migration file contents. If a migration has  been altered, the migrations will fail. This is to ensure consistency as migrations  are added."

See tests for more examples.

## Related Issues
Closes TRI-7

## Submission Checklist
- [x] Commits include the JIRA issue and the `#review` hashtag (e.g. `TRI-### #review <comment>`)
- [x] Code passes linter rules (`clj-kondo --lint src`)

## Testing

Run `clj -M:test`

### CODE REVIEW

File: test/triangulum/security_test.clj [hunk 1]
```diff
@@ -0,0 +1,28 @@
+(ns triangulum.security-test
+  (:require [clojure.test        :refer [is deftest testing use-fixtures]]
+            [clojure.java.io     :as io]
+            [triangulum.security :refer [hash-str hash-file]]
+            [triangulum.utils     :refer [delete-recursively]]))
```
Comment (Oliver-BE) [inline test/triangulum/security_test.clj:5]: Remove the extra space before `:refer` here.

File: src/triangulum/build_db.clj [hunk 2]
```diff
@@ -159,14 +160,16 @@
...
-               :requires    [:file]}})
+  {:backup        {:description "Create a .dump backup file using pg_dump."
+                   :requires    [:dbname :file]}
+   :build-all     {:description "Build / rebuild the entire data base."
+                   :requires    [:dbname]}
+   :functions     {:description "Build / rebuild all functions."
+                   :requires    [:dbname]}
+   :restore       {:description "Restore a database from a .dump file created by pg_dump."
+                   :requires    [:file]}
+   :apply-changes {:description "Applies the migration files under `src/sql/changes` in chronological order."
```
Comment (Oliver-BE) [inline src/triangulum/build_db.clj:171]: Why is this `:apply-changes` and not `:migrate`?

**Comment by sirmspencer** [comment]
Check your comments https://www.gnu.org/software/emacs/manual/html_node/elisp/Comment-Tips.html

What is ^:unit for?

src/triangulum/security.clj
9, Its nice to short hand variable names in dev.  We should all have auto complete now, so spell it out `hash-algorithm`. 
10, not quite sure on this one.  In Java if you declare an input type as string it will fail on call (compile time) if the value isnt a string.  Do you need to check again that its a sting if you are already type hinting?
13, do you need separate lets?  Java classes are not immutable so it might be possible to update md after its associated with digest. Or in a more clojury way, can you thread all of these into a single binding at the end?
16, same as 10, do we need these checks?  If so is it more clojury to add :pre?

src/triangulum/utils.clj
150, this seems more complicated that the solution from carbon tool.  file_io.clj, line 238:
```
(doseq [file (reverse (file-seq (io/file tmp-dir folder-name)))]
    (io/delete-file file))
```
Did I miss something that it does differently?

test/triangulum/security_test.clj
9, put the `(.exists (io/file "tmp")` check in the function if its needed.
17, what are you testing passing in a random string that already looks hashed?

src/triangulum/migrate.clj
61, Seems odd to mix str and format
68, Im not seeing the value in this binding.  I think you can rewrite the entire function into a single try catch.  If verbose, also post a message from the catch block.
78/81, can you rewrite this with a single when-let?
113, This call doesnt rely on anything from the let block.  Can you just call this before the let?
119, I dont think you need this when.  (some? (first new-changes)) will be nil if count is 0.
123, Are you sure this will work?  I usually find that dereffing in a if-let is always truthy.

**Comment by rjsheperd** [comment]
> What is ^:unit for?

Having the `^:unit` tag is to separate the tests which require a database (`^:db`).  To run all of the tests (locally), you can run `clj -M:test:test-db`.

**Comment by lambdatronic** [comment]
# src/triangulum/build_db.clj
- 2: `[java.io File]` -> `java.io.File`
- 163-172: Too much space after map keys.

# src/triangulum/migrate.clj
- 26: `(io/make-parents *migrations-dir* "dummy.txt")`
- 33-34: Add `^File` to prevent reflection.
- 39: Use `not=`
- 46: Wrap `nil-on-error` around `file-changed?` call.
- 60: "...after:\n- "
- 61: Use this simpler expression:
  ```clojure
  (str/join "\n-" (rest (drop-while #(not= file-name %) new-changes)))
  ```
- 67: Use `migration-path`
- 81: `uuid_generate_v4 ()` -> `uuid_generate_v4()`
- 82: `VARCHAR`             -> `varchar`
- 83: `VARCHAR`             -> `varchar`
- 84: `TIMESTAMP`           -> `timestamp`
- 116: I don't see how `:error` gets onto the `result` map. It looks
  like any errors in the SQL code will throw an `Exception` from
  `apply-migration` that isn't caught.

# src/triangulum/security.clj
- 2: Don't require `clojure.java.io`.
- 3: `[java.security MessageDigest]` -> `java.security.MessageDigest`
- 3: Move import above require.
- 11: Add ^String to `input` to prevent reflection.
- 17: No need to call `io/file`.

# src/triangulum/systemd.clj
- 44: Add `user` back to fn args.

# src/triangulum/utils.clj
- 9-13: Remove `nil-on-error` and import it from `triangulum.errors`.
- 150-159: This can be done more simply like so:
  ```clojure
  (defn delete-recursively [dir]
    (run! io/delete-file (reverse (file-seq (io/file dir)))))
  ```

# test/triangulum/migrate_test.clj
- 3: `clojure.java.io` comes before `clojure.test`.
- 48: End `DROP ROLE` statement with semicolon. and split over two lines to match `create-db` above.
- 51,55: You have to call `.close` on `con` or risk a resource leak.
- 77: Use lowercase for SQL types.
- 95: Use lowercase for SQL types.
- 98: `doseq [_ (range 0 5)]` -> `dotimes [_ 5]`
- 116: `( id` -> `(id`
- 116: Use lowercase for SQL types.
- 132: Use lowercase for SQL types.

# test/triangulum/security_test.clj
- 3: `clojure.java.io` comes before `clojure.test`.

**Comment by lambdatronic** [comment]
# deps.edn
- 44: Fix `:sha` indentation.

# src/triangulum/migrate.clj
- 36: Add `^File` to prevent reflection.
- 62,63: "\n-" -> "\n- "

# src/triangulum/systemd.clj
- 73,79: `:user` option is never used.

# src/triangulum/utils.clj
- 8,9: Keep libspecs in alphabetical order.

# test/triangulum/migrate_test.clj
- 78-79: Fix indentation.
- 84: SERIAL -> serial
- 102: Use lowercase for SQL types.
- 113-114: Fix indentation.
- 123: SERIAL -> serial
- 132-133: Fix indentation.
- 139: SERIAL -> serial

**Pull request merged**